src/Utils.cc

```diff
@@ -563,4 +574,275 @@ thread_type_t fbgemmGetThreadPartition(
   return th_info;
 }
 
+namespace {
+
+// implementation taken from pytorch/c10/util/llvmMathExtras.h
+template <typename T>
+size_t count_leading_zeros(T val) {
+  if (!val)
+    return std::numeric_limits<T>::digits;
+
+  // Use bisection method
+  size_t zero_bits = 0;
+  for (auto shift = std::numeric_limits<T>::digits >> 1; shift; shift >>= 1) {
+    T tmp = val >> shift;
+    if (tmp)
+      val = tmp;
+    else
+      zero_bits |= shift;
+  }
+  return zero_bits;
+}
+
+// histogram size per thread
+constexpr int RDX_HIST_SIZE = 256;
+
+void update_prefsum_and_offset_in_range(
+    int &offset,
+    const int bins_beg,
+    const int bins_end,
+    const int nthreads,
+    const int* const histogram,
+    int* const histogram_ps) {
+  for (int bins = bins_beg; bins < bins_end; ++bins) {
+    for (int t = 0; t < nthreads; ++t) {
+      histogram_ps[t * RDX_HIST_SIZE + bins] = offset;
+      offset += histogram[t * RDX_HIST_SIZE + bins];
+    }
+  }
+}
+
+void combine_prefix_sum(
+    const int nthreads,
+    const int elements_count,
+    const int* const histogram,
+    int* const histogram_ps) {
+  int offset = 0;
+  update_prefsum_and_offset_in_range(
+      offset, 0, RDX_HIST_SIZE, nthreads, histogram, histogram_ps);
+  histogram_ps[RDX_HIST_SIZE * nthreads] = offset;
+  // TODO(DamianSzwichtenberg): Is assert sufficient? In most cases, it will
+  // work only in debug build.
+  assert(offset == elements_count);
+  // Suppress unused variable warning
+  (void)elements_count;
+}
+
+void combine_prefix_sum_for_msb(
+    const int nthreads,
+    const int elements_count,
+    const int* const histogram,
+    int* const histogram_ps) {
+  int offset = 0;
+  update_prefsum_and_offset_in_range(
+      offset, 128, RDX_HIST_SIZE, nthreads, histogram, histogram_ps);
+  update_prefsum_and_offset_in_range(
+      offset, 0, 128, nthreads, histogram, histogram_ps);
+  histogram_ps[RDX_HIST_SIZE * (nthreads - 1) + 127] = offset;
+  // TODO(DamianSzwichtenberg): Is assert sufficient? In most cases, it will
+  // work only in debug build.
+  assert(offset == elements_count);
+  // Suppress unused variable warning
+  (void)elements_count;
+}
+
+template <typename K, typename V>
+void radix_sort_kernel(
+    const K* const input_keys,
+    const V* const input_values,
+    K* const output_keys,
+    V* const output_values,
+    const int elements_count,
+    int* const histogram,
+    int* const histogram_ps,
+    const int pass,
+    const bool pass_with_sign_bit = false) {
+  const auto tid = omp_get_thread_num();
+  const auto nthreads = omp_get_num_threads();
+  const auto elements_count_4 = elements_count / 4 * 4;
+
+  int* const local_histogram = &histogram[RDX_HIST_SIZE * tid];
+  int* const local_histogram_ps = &histogram_ps[RDX_HIST_SIZE * tid];
+
+  // Step 1: compute histogram
+  for (int i = 0; i < RDX_HIST_SIZE; i++) {
```

Comment:
Use `for (const auto i : c10::irange(RDX_HIST_SIZE)` if `c10::irange` is available. Please do this for all for-loops.

Reply:
I don't think I can do this as FBGEMM CPU does not have PyTorch in its dependencies.